python/src/main/java/org/apache/apex/malhar/python/base/ApexPythonEngine.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.python.base;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+
+import org.apache.apex.malhar.python.base.requestresponse.PythonInterpreterRequest;
+import org.apache.apex.malhar.python.base.requestresponse.PythonRequestResponse;
+
+/**
+ * Defines the methods that needs to be implemented by the Python operator implementations. The first implementation
+ *  takes the approach of in-memory interpreter using JEP. Other possibilities are using Py4J which is an inter process
+ *   communication model.
+ */
+public interface ApexPythonEngine
+{
+  /***
+   * Used to perform any pre interpreter processing.
+   * @param preInitConfigs The configuration that is going to be used by the interpreter
+   * @throws ApexPythonInterpreterException if there is an issue in executing the pre interpreter logic
+   */
+  void preInitInterpreter(Map<PythonInterpreterConfig,Object> preInitConfigs) throws ApexPythonInterpreterException;
+
+  /***
+   * Starts the interpreter.
+   * @throws ApexPythonInterpreterException if library not locatable or any other issue starting the interpreter
+   */
+  void startInterpreter() throws ApexPythonInterpreterException;
+
+  /***
+   * Used to perform any logic that needs to be executed after the interpreter is started but before any tuples start
+   *  getting processed. Example, setting the starting state of the variables that are used in tuple processing.
+   * @throws ApexPythonInterpreterException
+   */
+  void postStartInterpreter() throws ApexPythonInterpreterException;
```

Comment:
why can such logic not be included in startInterpreter?

Reply:
Looking at the implementation of setup, I see the following:
```
     apexPythonEngine.preInitInterpreter(preInitConfigurations);
     apexPythonEngine.startInterpreter();
     apexPythonEngine.postStartInterpreter();
```
If the operator has these calls as a single logical unit, then why are they broken out in the interpreter interface? I understand that interpreter internal the differentiation makes sense, just wondering if that means it should also be visible in the interface. Is that motivated by usage outside of an Apex operator?

Reply:
The primary motivation was for support beyond the methods of an Apex operator ( example as a utility usage etc). Bundling all of these in the single setup() method does not seem a good idea as you pointed out. 

Moved the postStartInterpreter() to another method which can be overridden if required that by default includes the postStartInterpreter() call. Operator implementors can make use this decoupled approach to only "touch" the post interpreter logic without worrying about setup() which now only deals with the interpreter boot up.